### PR TITLE
#164829378 Implement Article Bookmark Feature

### PR DESCRIPTION
**Pull request by Kyppy**
#### Description
This PR creates and implements an article bookmarking feature. It enables users to create a convenient reference to a particular article. Users can then select an article from their list of saved bookmarks.
The is PR implements the following endpoints:

|  Method  | Endpoint |  Description |
|---|---|---|
| POST  |  localhost/api/bookmark/:article_id | Creates a bookmark for an article. |
| GET  | localhost/api/bookmarks/| Returns a list of user-specific bookmarks. |

#### Description of Task to be completed?
An authenticated user should be able to bookmark articles for reading later.
An authenticated user should be able to manage those they follow and be able to see those who follow them.

*NOTE*
Authenticating as a user requires you to use the sign-up and login views.
Please ensure you use an email address that you have access to. A verification email will be sent that you will have to use in order to complete the sign-up process.

1) Enter Sign-up Details
Access the endpoint using: http://127.0.0.1:8000/api/users/
Example sign-up payload:
```
{
  "user": {
    "email": "demo@mail.com",
    "username" :"Mary",
    "password":abc123ok
  }
}
```
2) Verify Sign-up Via Email
After you have posted your sign-up details an email will be sent to the email address you provided. Be aware that it will most likely appear in your spam folder. Open the email and click on the "Confirm Email Address" button. You will be redirected to an empty Heroku page. This is expected. Using this page, you will then copy everything in the URL __after__  "https://ah-the-answer-backend-st-pr-21.herokuapp.com/api/" (see example below)

Screen Shot 2019-04-09 at 14 58 05

Then with that URL create a GET request using the following URL (see example below):
http://127.0.0.1:8000/api/<copy paste the Heroku URL here>

Screen Shot 2019-04-09 at 15 05 51

You will receive a message confirming your verification.

3) Log In
Access the endpoint using: http://127.0.0.1:8000/api/users/login/
Example login payload:
```
{
  "user": {
    "email": "demo@mail.com",
    "password":abc123ok
  }
}
```
After a successful login, a token will be returned. Copy and include it in the authorization headers when attempting to access any of the 'follow' views excluding the view that counts and returns the number of 'follows'/'followings' for a user.

4) Create An Article
Access the create article endpoint using:  http://127.0.0.1:8000/api/articles/
Example 'article' payload:
```
{
    "article":
        {
            "title": "Birds And Nerds",
            "body": "Coding Poetry",
            "description": "Pleasant Words",
            "is_published": true
        }
}
```
#### How should this be manually tested?
* Clone repo to local machine
* 'Cd' into cloned repository folder
* Create a virtual environment by running `virtualenv venv`
* Activate the virtualenv `source venv\bin\activate`
* Install dependencies `pip install -r dev-requirements.txt`
* Run the Django server to serve up the app `python manage.py runserver`

###### Test with Postman using the payloads
To test these endpoints JSON payloads are not necessary. All that is required is the creation of a valid user, valid article and to include their login token in the authorization headers of each request. All input data is sent through the URL.

For example, to create a bookmark for an article with an id of 10 you would make the following POST request:
http://127.0.0.1:8000/api/bookmark/10/

#### How has this been tested?
- [x] Unittest
Run 'python manage.py test' to run the test and receive a report with test coverage.

#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/164829378

### CODE REVIEW

**Comment by LarryKarani** [comment]
Recommendable job. My only concern is that when you try to bookmark an article that shares the same title with another article you get the following error
 ```
{
    "error": "No article with that title found."
}
```
to fix this you can maybe use the article slag to uniquely identify the article instead of the title.

File: authors/apps/bookmark/views.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        user object and the article bookmark."""
+        user = self.request.user
+        username = user.username
+        try:
+            article = Article.objects.get(title=article_title)
+            article_id = article.pk
+        except:
+            return Response({"error": "No article with that title found."},
+                            status=status.HTTP_404_NOT_FOUND)
+        bookmark = Bookmark.objects.filter(article_title=article_title).filter(
```
Comment (kirega) [inline authors/apps/bookmark/views.py]: I believe the chained filter can be refactored to  `Bookmark.objects.filter(article_title=article_title,article_id=article_id)`. Are there any particular reasons you chose to do it like this?

File: authors/apps/bookmark/views.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+from rest_framework.views import APIView
+from rest_framework.exceptions import APIException
+
+from ..articles.models import Article
+from ..bookmark.models import Bookmark
+from ..authentication.models import User
+
+
+class CreateBookmark(APIView):
+    """This class allows contains views that allow a user to
```
Comment (Raywire) [inline authors/apps/bookmark/views.py]: Good work on the comprehensive docstrings. Please fix the grammatical error here.

**Review by LarryKarani — APPROVED** [review]

**Comment by kirega** [comment]
Feedback Implemented.  I believe it is now ready for deployment.

**Review by kirega — APPROVED** [review]

**Pull request merged**